keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,894 @@
+driver deployments.
+
+### API
+
+All Kubernetes API changes are under the `CSIStorageCapacity` feature
+gate.
+
+#### CSIStoragePool
+
+The prefix “CSI” is used to avoid confusion with other types that have
+a similar name and because some aspects are specific to CSI.
+
+```
+// CSIStoragePool identifies one particular storage pool and
+// stores the corresponding attributes. The spec is read-only.
+//
+// There should never be two objects with the same spec. If
+// a consumer of this information finds two objects with the
+// same spec, it should use the information from the one with the
+// most recent CreationTimestamp.
+type CSIStoragePool struct {
+	metav1.TypeMeta
+	// Standard object's metadata. The name has no particular meaning and just has to
+	// meet the usual requirements (length, characters, unique). To ensure that
+    // there are no conflicts with other CSI drivers on the cluster, the recommendation
+    // is to use sp-<uuid>.
+	//
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta
+
+	Spec   CSIStoragePoolSpec
+	Status CSIStoragePoolStatus
+}
+
+// CSIStoragePoolSpec contains the constant attributes of a CSIStoragePool.
+//
+// A pool might only be accessible from a subset of the nodes in the
+// cluster. That subset can be identified either via NodeTopology or
+// NodeList, but not both. If neither is set, the pool is assumed
+// to be available in the entire cluster.
+type CSIStoragePoolSpec struct {
+    // The CSI driver that provides access to the storage pool.
+	// This must be the string returned by the CSI GetPluginName() call.
+	DriverName string
+
+	// NodeTopology can be used to describe a storage pool that is available
+	// only for nodes matching certain criteria.
+	// +optional
+	NodeTopology *v1.NodeSelector
+
+	// Nodes can be used to describe a storage pool that is available
+	// only for certain nodes in the cluster.
+	//
+	// +listType=set
+	// +optional
+	Nodes []string
+}
+
+// CSIStoragePoolStatus contains runtime information about a CSIStoragePool.
+//
+// It is expected to be extended with other
+// attributes which do not depend on the storage class, like health of
+// the pool. Therefore it has the list of
+// `CSIStorageByClass` instances instead of just the capacity
+// and the storage class being in the spec.
+type CSIStoragePoolStatus struct {
+	// Some information, like the actual usable capacity, may
+	// depend on the storage class used for volumes.
+	//
+	// +patchMergeKey=storageClassName
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=storageClassName
+	// +optional
+	Classes []CSIStorageByClass `patchStrategy:"merge" patchMergeKey:"storageClassName" json:"classes,omitempty" protobuf:"bytes,4,opt,name=classes"`
+}
+
+// CSIStorageByClass contains information that applies to one storage
+// pool of a CSI driver when using a certain storage class.
+type CSIStorageByClass struct {
```

Comment:
> So it's better that each CSI driver defines CSIStorages not corresponding to StorageClass. For  example, for "Example: affect of storage classes", define CSIStorages for striped and mirrored. As a side note, this can remove <fallback>.

But there's still a `CSIStorage.storageClass` string field? If not, then how would the Kubernetes scheduler look up capacity information for a PVC? All it has are:
- storage class name
- CSI driver name inside the storage class
- several parameters, some of them standardize (`fsType`), some not (opaque key/value pairs)

As long as there's still a `CSIStorage.storageClass`, then I agree that `CSIStorageByClass` could be eliminated by creating multiple `CSIStorage` instances. The reason for not proposing that solution is that @cdickmann pointed out the need to also represent other information about storage which isn't dependent on the storage class, like health. Such information nicely fits into the current proposal as a future extension, but would be awkward when there are multiple `CSIStorage` instances that all refer to the same underlying storage.

Reply:
OK, I see.